File: web/di_core/DataRelationship/Domain/QueryView.ts
import { InlineSqlView, SqlQuery, TableSchema } from '@core/domain';

export enum QueryViews {
  Table = 'table_view',
  SQL = 'sql_view'
}

export abstract class QueryView {
  abstract className: QueryViews;
  abstract aliasName: string;

  static fromObject(obj: QueryView) {
    switch (obj.className) {
      case QueryViews.Table: {
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        return new TableView((obj as TableView).dbName, (obj as TableView).tblName, obj.aliasName);
      }
      case QueryViews.SQL:
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        return new InlineSqlView(obj.aliasName, (obj as InlineSqlView).query);
    }
  }

  static default() {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new TableView('', '', '');
  }

  static isTableView(view: QueryView) {
    return view.className === QueryViews.Table;
  }
}

export class TableView implements QueryView {
  className = QueryViews.Table;
  constructor(public dbName: string, public tblName: string, public aliasName: string) {}

  static fromObject(obj: TableView) {
    return new TableView(obj.dbName, obj.tblName, obj.aliasName);
  }
}

File: rocket-bi-server/src/main/scala/co/datainsider/bi/domain/query/QueryView.scala
package co.datainsider.bi.domain.query

import co.datainsider.bi.util.StringUtils
import com.fasterxml.jackson.annotation.JsonSubTypes.Type
import com.fasterxml.jackson.annotation.{JsonSubTypes, JsonTypeInfo}

/**
  * a view is a data table in which holds data to be selected and applied calculated functions (select, group by, sum, min, max...)
  * a view contains multiple fields and describes the source of data (from real data table, or from result of an sql...)
  */
@JsonTypeInfo(
  use = JsonTypeInfo.Id.NAME,
  include = JsonTypeInfo.As.PROPERTY,
  property = "class_name"
)
@JsonSubTypes(
  Array(
    new Type(value = classOf[TableView], name = "table_view"),
    new Type(value = classOf[SqlView], name = "sql_view")
  )
)
abstract class QueryView {
  val aliasName: String
}

case class SqlView(aliasName: String, query: Query) extends QueryView

case class TableView(dbName: String, tblName: String, aliasViewName: Option[String] = None) extends QueryView {
  val fullName: String = s"$dbName.$tblName"

  override val aliasName: String = aliasViewName.getOrElse("tbl_" + StringUtils.shortMd5(fullName))
}
